pkg/migration/migration.go

```diff
@@ -0,0 +1,259 @@
+	// things intended for the specified version.
+	//
+	// It's important to note that the closure is being executed in the context of a
+	// distributed transaction that may be automatically retried. So something like
+	// the following is an anti-pattern:
+	//
+	//     processed := 0
+	//     _ = h.IterateRangeDescriptors(...,
+	//         func(descriptors ...roachpb.RangeDescriptor) error {
+	//             processed += len(descriptors) // we'll over count if retried
+	//             log.Infof(ctx, "processed %d ranges", processed)
+	//         },
+	//     )
+	//
+	// Instead we allow callers to pass in a callback to signal on every attempt
+	// (including the first). This lets us salvage the example above:
+	//
+	//     var processed int
+	//     init := func() { processed = 0 }
+	//     _ = h.IterateRangeDescriptors(..., init,
+	//         func(descriptors ...roachpb.RangeDescriptor) error {
+	//             processed += len(descriptors)
+	//             log.Infof(ctx, "processed %d ranges", processed)
+	//         },
+	//     )
+	//
+	// [1]: pkg/kv/kvserver/batch_eval/cmd_migrate.go
+	IterateRangeDescriptors(ctx context.Context, size int, init func(), f func(descriptors ...roachpb.RangeDescriptor) error) error
+}
+
+// Migration defines a program to be executed once every node in the cluster is
+// (a) running a specific binary version, and (b) has completed all prior
+// migrations.
+//
+// Each migration is associated with a specific internal cluster version and is
+// idempotent in nature. When setting the cluster version (via `SET CLUSTER
+// SETTING version`), the manager process determines the set of migrations
+// needed to bridge the gap between the current active cluster version, and the
+// target one. See [1] for where that happens.
+//
+// To introduce a migration, start by adding version key to pkg/clusterversion
+// and introducing a corresponding internal cluster version for it. See [2] for
+// more details. Following that, define a Migration in the migrations package
+// and add it to the appropriate migrations slice to the registry. Be sure to
+// key it in with the new cluster version we just added. During cluster
+// upgrades, once the operator is able to set a cluster version setting that's
+// past the version that was introduced (typically the major release version
+// the migration was introduced in), the manager will execute the defined
+// migration before letting the upgrade finalize.
+//
+// If the migration requires below-Raft level changes ([3] is one example),
+// you'll need to add a version switch and the relevant KV-level migration in
+// [4]. See IterateRangeDescriptors and the Migrate KV request for more details.
+//
+// [1]: `(*Manager).Migrate`
+// [2]: pkg/clusterversion/cockroach_versions.go
+// [3]: truncatedStateMigration
+// [4]: pkg/kv/kvserver/batch_eval/cmd_migrate.go
+//
+type Migration interface {
+	ClusterVersion() clusterversion.ClusterVersion
```

Comment:
Given that we want only two real types to implement this, KVMigration and SQLMigration, should we define a hidden method here to restrict the list of types implicitly implementing the Migration interface?

Reply:
Done.